### PR TITLE
Initial Commit of dagster-cloud.yaml doc

### PR DESCRIPTION
**Pull request by odette-elementl**

### CODE REVIEW

**Comment by odette-elementl** [comment]
Do we need separate code file when its yaml? @erinkcochran87

File: docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+Parameter reference table:
+
+| Parameter                            | Description                                                                           | Data type            |
+| ------------------------------------ | ------------------------------------------------------------------------------------- | -------------------- |
+| `locations.location_name`            | The name of your code location that will appear in the Dagster UI code locations page | string               |
+| `locations.code_source.package_name` | The folder that contains your dagster code                                            | string (folder name) |
+| `locations.code_source.python_file`  | The python file that contains your dagster code                                       | string (.py file)    |
+| `locations.code_source.module_name`  | The module that contains your dagster code                                            | string (module name) |
+
+Single code location:
```
Comment (slopp) [inline docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx:42]:
```suggestion
Example for a single code location:
```

I'd add the example pre-amble throughout

File: docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+├── quickstart_etl_tests
+├── dagster_cloud.yaml
+├── pyproject.toml
+├── setup.cfg
+└── setup.py
+```
+
+## Code Source
+
+The only **mandatory** section in the dagster_cloud.yaml file is setting up your code locations. If you have multiple code locations, you will have multiple code locations within a single dagster_cloud.yaml file. Dagster code can be sourced from a package, a python file, or a module.
```
Comment (slopp) [inline docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx:31]:
```suggestion
The only **mandatory** section in the dagster_cloud.yaml file is setting up your code locations. If you have multiple code locations, you will have multiple code locations within a single dagster_cloud.yaml file. Dagster code can be sourced from a package, a python file, or a module, and one of these parameters should be specified for each code location.
```

File: docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+Parameter reference table:
+
+| Parameter                            | Description                                                                           | Data type            |
+| ------------------------------------ | ------------------------------------------------------------------------------------- | -------------------- |
+| `locations.location_name`            | The name of your code location that will appear in the Dagster UI code locations page | string               |
+| `locations.code_source.package_name` | The folder that contains your dagster code                                            | string (folder name) |
```
Comment (slopp) [inline docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx:38]:
```suggestion
| `locations.code_source.package_name` |  Use if your dagster code is structured as a package | string (folder name) |
```

File: docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+| `locations.code_source.python_file`  | The python file that contains your dagster code                                       | string (.py file)    |
```
Comment (slopp) [inline docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx:39]:
```suggestion
| `locations.code_source.python_file`  | Use if your dagster code is in a single file | string (.py file)    |
```

File: docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+Single code location:
+
+```yaml
+#dagster_cloud.yaml
+
+locations:
+  - location_name: example_location
+    code_source:
+      package_name: my_dagster_project
+    build:
```
Comment (slopp) [inline docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx:51]: We should remove the build section here because it is not explained until later in the doc.

File: docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+locations:
+  - location_name: data-eng-pipeline
+    code_source:
+      package_name: hooli_data_eng
+    working_directory: ./hooli_data_eng
+```
+
+## Build: Docker Registry
+
```
Comment (slopp) [inline docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx:91]: We should add a pre-amble here.

If you are using Dagster Cloud Hybrid with a Docker, Kubernetes, or ECR agent you will need to tell Dagster Cloud where the Docker image containing your code is stored. This is done through the`build` section of dagster_cloud.yaml.

File: docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+Parameter reference table:
+
+| Parameter                   | Description                                                                                                                                                       | Data type                |
+| --------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------ |
+| `locations.build.directory` | The directory in your project code that you want to deploy, the default is `.`. If you have sub-folders you can specific the path to only deploy that sub-project | string (path)            |
```
Comment (slopp) [inline docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx:96]:
```suggestion
| `locations.build.directory` | Use if you are using the hybrid cloud deploy GitHub action that builds a Docker image for you. Supply the directory in your project code that you want to include in the image, the default is `.`. If you have sub-folders you can specify the path to only include that sub-directory | string (relative path)            |
```

File: docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx (hunk 6 above)
Comment (slopp) [reply docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx:91]: @gibsondan is this all true? I'm unclear as to what situations require `build` vs just including the `image` value. I think it has to do with whether you use our build_and_deploy action or just deploy... Any ideas on what language could make that distinction clear here?

File: docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+```yaml
+#dagster_cloud.yaml
+
+locations:
+  - location_name: example_location
+    code_source:
+      package_name: my_dagster_project
+    build:
+      directory: ./
+      registry: 764506304434.dkr.ecr.us-west-2.amazonaws.com/branch-deployments-gh-action-test
```
Comment (slopp) [inline docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx:110]: We should make this registry reference, and all subsequent, more clearly a thing the customer needs to provide to avoid them copying and pasting this value

Maybe `registry: your-docker-image-registry/image-name`

File: docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+```
+
+Multiple code locations:
```
Comment (slopp) [inline docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx:113]: I think this is not needed since we covered how to do multiple code locations earlier, and this section just focuses on the new build section

File: docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+| --------------------------- | -------------------------------------- | ------------- |
+| `locations.executable_path` | The location of your python executable | string (path) |
+
+```yaml
+#dagster_cloud.yaml
+
+locations:
+  - location_name: data-eng-pipeline
+    code_source:
+      package_name: hooli_data_eng
```
Comment (slopp) [inline docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx:149]: location name and package name could reflect the executable path, eg data-eng stuff vs ml stuff

File: docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+    code_source:
+      python_file: definitions.py
+    build:
+      directory: ./hooli_basics
+      registry: 764506304434.dkr.ecr.us-west-2.amazonaws.com/hooli-data-science-prod
+```
+
+## Python Executable
+
+By default, Dagster Cloud will use the python executable that is installed on the agent. If you want to use a different python executable, you can specify it in the dagster_cloud.yaml file. It might also be useful to have different python executables for different code locations.
```
Comment (slopp) [inline docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx:135]:
```suggestion
If you are using Dagster Cloud Serverless, the Python executable to use is specified [here](https://docs.dagster.io/dagster-cloud/deployment/serverless#using-a-different-python-version). For Dagster Cloud Hybrid, the python executable that is installed globally in the image, or the default Python executable on the local system if you use the local agent, will be used. If you want to use a different python executable, you can specify it in the dagster_cloud.yaml file. It might also be useful to have different python executables for different code locations.
```

File: docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+      package_name: hooli_data_eng
+    executable_path: venvs/path/to/dataengineering_spark_team/bin/python
+  - location_name: basics
+    code_source:
+      python_file: definitions.py
+    executable_path: venvs/path/to/ml_tensorflow/bin/python
+```
+
+## Other Configurations
+
```
Comment (slopp) [inline docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx:158]: I would ditch the section on secrets here. It is true that you can set secrets in some Dagster Cloud Hybrid setups, but many customers will still use the Cloud environment variable workflow. Instead I would say this: 

In addition to these settings, different Dagster Cloud Hybrid agents support additional configuration options. To see details: 

- [ECS](https://docs.dagster.io/dagster-cloud/deployment/agents/amazon-ecs/configuration-reference#per-location-configuration)
- [Kubernetes](https://docs.dagster.io/dagster-cloud/deployment/agents/kubernetes/configuration-reference#per-location-configuration)
- [Docker](https://docs.dagster.io/dagster-cloud/deployment/agents/docker/configuration-reference#environment-variables-and-secrets)

**Review by slopp — CHANGES_REQUESTED** [review]
Thanks for taking this on... it is definitely a big and murky task!

I left one large suggestion and then smaller ones.

I also had a question for Daniel on the `build` section.

File: docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+| `working_directory` | The path of the directory that Dagster should run from | string (path) |
+
+```yaml
+#dagster_cloud.yaml
+
+locations:
+  - location_name: data-eng-pipeline
+    code_source:
+      package_name: hooli_data_eng
+    working_directory: ./hooli_data_eng
```
Comment (shalabhc) [inline docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx:87]: I suggest a different directory name because this field is unrelated to `package_name`.

File: docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx (hunk 6 above)
Comment (shalabhc) [reply docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx:91]: This section should just be called the "Build directory". You can use a `build: directory:` in serverless as well. It works for both docker and pex. Basically it specifies which directory contains your setup.py or requirements.txt file for this code location.

Comment (shalabhc) [reply docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx:91]: We should say this is useful if your setup.py or requirements.txt is not at your project root but in a subdirectory. "This is common if you have multiple python packages within a single dagster project"

File: docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx (hunk 7 above)
Comment (shalabhc) [reply docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx:96]: This doesn't apply only to hybrid, see my comment above.

File: docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+```
+
+## Build: Docker Registry
+
+Parameter reference table:
+
+| Parameter                   | Description                                                                                                                                                       | Data type                |
+| --------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------ |
+| `locations.build.directory` | The directory in your project code that you want to deploy, the default is `.`. If you have sub-folders you can specific the path to only deploy that sub-project | string (path)            |
+| `locations.build.registry`  | The Docker registry to push your code location to in the registry                                                                                                 | string (docker registry) |
```
[the remark this answers is not in the document]
Comment (shalabhc) [reply docs/content/dagster-cloud/developing-testing/dagster-cloud-yaml.mdx:97]: Registry is hybrid only.

**Review by shalabhc — CHANGES_REQUESTED** [review]

**Comment by odette-elementl** [comment]
messed up rebase
